File: go-skvdb-client/src/storage/table.go
package storage

import (
	"fmt"
	"go-skvdb-client/src/dto"
	"go-skvdb-client/src/service/networkService"
	"strconv"
)

type Table interface {
	Set(k string, v string) error
	Get(k string) (string, error)
	Delete(k string) error
	ContainsKey(k string) (bool, error)
	GetTableMetaData() TableMetaData
}

type TableImpl struct {
	username string
	password string

	networkService *networkService.NetworkService
	TableMetaData  TableMetaData
}

func newTableImpl(networkService *networkService.NetworkService, username string, password string, tableName string) *TableImpl {
	return &TableImpl{networkService: networkService, username: username, password: password, TableMetaData: TableMetaData{Name: tableName}}
}

func (t *TableImpl) Set(k string, v string) error {
	const op = "storage.table.Set"

	body := make(map[string]string)
	body["key"] = k
	body["value"] = v
	body["table"] = t.TableMetaData.Name
	response, err := t.networkService.Send(dto.Request{
		Username:   t.username,
		Password:   t.password,
		MethodName: "set",
		Body:       body,
	})
	if err != nil {
		return fmt.Errorf("%s: %w", op, err)
	}
	if response.Result == "OK" {
		return nil
	}
	return fmt.Errorf("%s", response.Error)
}

func (t *TableImpl) Get(k string) (string, error) {
	const op = "storage.table.Get"

	body := make(map[string]string)
	body["key"] = k
	body["table"] = t.TableMetaData.Name
	response, err := t.networkService.Send(dto.Request{
		Username:   t.username,
		Password:   t.password,
		MethodName: "get",
		Body:       body,
	})
	if err != nil {
		return "", fmt.Errorf("%s: %w", op, err)
	}
	if response.Result == "OK" {
		return response.Body["value"], nil
	}
	return "", fmt.Errorf("%s", response.Error)
}

func (t *TableImpl) Delete(k string) error {
	const op = "storage.table.Delete"

	body := make(map[string]string)
	body["key"] = k
	body["table"] = t.TableMetaData.Name
	response, err := t.networkService.Send(dto.Request{
		Username:   t.username,
		Password:   t.password,
		MethodName: "delete",
		Body:       body,
	})
	if err != nil {
		return fmt.Errorf("%s: %w", op, err)
	}
	if response.Result == "OK" {
		return nil
	}
	return fmt.Errorf("%s", response.Error)
}

func (t *TableImpl) ContainsKey(k string) (bool, error) {
	const op = "storage.table.ContainsKey"

	body := make(map[string]string)
	body["key"] = k
	body["table"] = t.TableMetaData.Name
	response, err := t.networkService.Send(dto.Request{
		Username:   t.username,
		Password:   t.password,
		MethodName: "contains_key",
		Body:       body,
	})
	if err != nil {
		return false, fmt.Errorf("%s: %w", op, err)
	}
	if response.Result == "OK" {
		return strconv.ParseBool(response.Body["result"])
	}
	return false, fmt.Errorf("%s", response.Error)
}

func (t *TableImpl) GetTableMetaData() TableMetaData {
	return t.TableMetaData
}

File: skvdb-client/src/main/java/org/skvdb/storage/Table.java
package org.skvdb.storage;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import org.skvdb.common.storage.TableMetaData;
import org.skvdb.dto.Request;
import org.skvdb.dto.RequestResult;
import org.skvdb.dto.Result;
import org.skvdb.network.NetworkService;

import java.util.HashMap;
import java.util.Map;

public class Table {
    private final NetworkService networkService;
    private final TableMetaData tableMetaData;

    private final String username;
    private final String password;

    public Table(NetworkService networkService, TableMetaData tableMetaData, String username, String password) {
        this.networkService = networkService;
        this.tableMetaData = tableMetaData;
        this.username = username;
        this.password = password;
    }

    public String get(String key) {
        Map<String, String> body = new HashMap<>();
        body.put("key", key);
        body.put("table", tableMetaData.name());
        Request request = new Request(username, password, null, "get", body);
        Result result = networkService.send(request);
        if (result.getRequestResult().equals(RequestResult.OK)) {
            return result.getBody().get("value");
        }
        throw new RuntimeException();
    }

    public void set(String key, String value) {
        Map<String, String> body = new HashMap<>();
        body.put("key", key);
        body.put("value", value);
        body.put("table", tableMetaData.name());
        Request request = new Request(username, password, null, "set", body);
        Result result = networkService.send(request);
        if (result.getRequestResult().equals(RequestResult.OK)) {
            return;
        }
        throw new RuntimeException();
    }

    public String convertToJson(String value) {
        ObjectWriter ow = new ObjectMapper().writer().withDefaultPrettyPrinter();
        try {
            return ow.writeValueAsString(value);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    public void delete(String key) {

    }

    public boolean containsKey(String key) {
        return false;
    }

    public TableMetaData getTableMetaData() {
        return tableMetaData;
    }
}
